perf(binary): move the WABinary codec to the Rust bridge

jlucaso1 opened the pull request:
Stacked on #2067. Review that one first; this branch was rewritten on top of it and the diff here is only the WABinary codec.

## Summary

`WABinary` encode and decode move to the Rust bridge. The tree crosses the boundary once, as a flat buffer, rather than field by field: pulling a node apart from Rust through `Reflect` was 56% of the encode profile, and returning a handle whose fields materialize on access moves the cost into the walk instead of removing it.

A decode leaves its four sections (interned strings, string offsets, a u32 layout stream, blob bytes) where the decoder wrote them in linear memory and publishes their offsets to a fixed address. JavaScript reads them through cached whole-memory views, so a decode allocates no typed arrays at all. Byte content is the one thing that outlives the call, so that section is copied once and the leaves are views into the copy.

## Measurement

Against `baileys@7.0.0-rc.9`, the pure TypeScript codec this replaces. Both decoders warmed first, process pinned to one core, on the stanzas the socket actually moves:

| stanza | encode (wasm / ts) | decode (wasm / ts) |
| --- | --- | --- |
| `ack`, 31 B | **0.73 us** / 1.09 us | **473 ns** / 572 ns |
| one to one message, 351 B | **1.63 us** / 4.62 us | **1.74 us** / 1.95 us |
| device fanout of 8, 1.4 KB | **6.35 us** / 19.44 us | **9.98 us** / 12.55 us |
| device fanout of 64, 8.4 KB | **38.9 us** / 137.3 us | **61.3 us** / 79.5 us |
| app state patch, 12.9 KB | **13.9 us** / 97.9 us | **6.66 us** / 8.76 us |

The warm-up matters and is worth flagging for anyone rerunning it: without it the first case measured absorbs the JIT warm-up for the whole file, about 150 ns, and it lands on whichever stanza happens to be first. That artifact read as the smallest stanza being the one case the bridge lost.

Memory is flat. The intern table that dedups strings within a decode is a fixed 512-slot direct-mapped table, so a socket that decodes two million stanzas with distinct message ids stays at 1.3 MB of linear memory. An earlier revision of this branch aged entries out by a round counter instead of removing them and reached 274 MB over the same run, which WASM never returns to the host.

## Wire parity

Byte for byte against rc.9 on the recorded stanzas in `test/parity.test.ts` and `test/server-response-parity.test.ts`, plus a re-encode of every decode reproducing the original frame.

Three behaviours were lost in an earlier revision and are restored here, each with a test:

- A jid with no user renders as a bare `s.whatsapp.net` in the core, and `handleEncryptNotification` routes on `from === S_WHATSAPP_NET`, so dropping the `@` sent the server's own pre-key count down the identity-change branch and replenishment never ran.
- Attributes set to `null` or `undefined` were stringified onto the wire as the literal text. They reach the encoder unset from real call sites, a USync user queried by phone and a media retry with no participant among them.
- A node with no tag was interned and shipped as a malformed stanza rather than throwing, and a child left out of a conditionally built list arrived as `null` and crashed the encoder rather than being dropped.

## Also in this diff, and separable if you would rather

Two things found while profiling the send path that are not the codec. Say the word and I will move them to their own PR:

- `promiseTimeout` captured `new Error().stack` and so did `delayCancellable`, which it calls. Capturing frames costs about 1.2 us; reading `.stack` makes V8 format them and costs about 11 us. It was not confined to the timeout path either, because settling ran `cancel`, which built a Boom and read its stack. Every outbound frame paid for two of those. Formatting is now deferred behind a getter and `promiseTimeout` goes from 22.0 us to 3.0 us.
- `generateParticipantHashV2` ran on every send and was discarded: `extraAttrs` had already been spread into the `<enc>` nodes by the two awaited `createParticipantNodes` calls above it. Fixing the order would not have helped, since the participant hash belongs on the message stanza rather than on `<enc>`, so it is removed rather than moved.

## Validation

```
pnpm --filter baileys test          # 594 passed
pnpm --filter whatsapp-rust-bridge test
cargo clippy --target wasm32-unknown-unknown --all-targets -- -D warnings
pnpm lint
```

The flat codec had no test of its own before this, only a bench. It gets one covering wire parity against a recorded stanza, each content kind, the token boundary, interning across decodes, and the two lifetime rules the borrowed sections depend on.

## Merge order

This needs #2067 first, and #2067 needs `whatsapp-rust-bridge` published with the flat exports: the tarball on npm today does not carry `decodeNodeFlat`, `encodeNodeFlat` or `tokenTable`, so a clean checkout cannot resolve them without building the bridge locally.

The previous state of this branch is kept at `jlucaso1/Baileys@backup/pr-1698-pre-rewrite` if anyone wants to read the earlier approach.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **Improvements**
  - Improved binary message encoding and decoding for faster, more efficient processing.
  - Enhanced handling of WhatsApp message content, including nested messages, binary data, and server-only JIDs.
  - Improved validation and resilience when processing malformed or unexpected binary data.
  - Added clearer timeout and cancellation behavior, including improved diagnostic stack information.

- **Bug Fixes**
  - Corrected preservation of server-only JID prefixes during message round trips.
  - Prevented unnecessary participant metadata from being added to outgoing messages.

- **Tests**
  - Expanded coverage for binary compatibility, timeout behavior, malformed input, and message round trips.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jlucaso1 commented on src/Types/Signal.ts:
```diff
@@ -14,12 +14,12 @@ type ProcessSenderKeyDistributionMessageOpts = {
 	jid: string
 	type: 'pkmsg' | 'msg'
-	ciphertext: Uint8Array
+	ciphertext: Buffer
```
This is needed because libsignal and WAGroup rely heavy on NodeJs Buffer (this will change in the libsignal rewrite too), but the new implementation of binary-wasm rely on pure Uint8Array.

jlucaso1 replied:
This can be more risky and there is the fact that the libsignal repository is not versioned correctly, which can lead to problems in older versions of baileys.

purpshell replied:
> This can be more risky and there is the fact that the libsignal repository is not versioned correctly, which can lead to problems in older versions of baileys.

That's correct, I failed to notice that. We'll make the changes then revert them.

purpshell commented:
Can you benchmark this? Mock test sendNode and pass through a thousand or so nodes and also make sure to make some different nodes chock-full of data and fields and test the difference.

purpshell commented on src/WABinary/jid-utils.ts:
```diff
@@ -1,4 +1,4 @@
-export const S_WHATSAPP_NET = '@s.whatsapp.net'
+export const S_WHATSAPP_NET = 's.whatsapp.net'
```
not an issue though (discussed with lucas)

purpshell commented on package.json:
```diff
@@ -47,6 +47,7 @@
     "music-metadata": "^11.7.0",
     "pino": "^9.6",
     "protobufjs": "^7.2.4",
+    "whatsapp-rust-bridge": "^0.1.3",
```
Repo will be transferred

jlucaso1 commented:
[bench.zip](https://github.com/user-attachments/files/21877729/bench.zip)
Benchmark results:
```bash
#using deprecated parameters for the initialization function; pass a single object instead # need to solve this warning later
clk: ~4.21 GHz
cpu: 12th Gen Intel(R) Core(TM) i5-12450H
runtime: node 22.18.0 (x64-linux)

benchmark                   avg (min … max) p75 / p99    (min … top 1%)
------------------------------------------- -------------------------------
• Encoding (JS Object -> Binary)
------------------------------------------- -------------------------------
Rust WASM (warmup-ignore)     45.61 µs/iter  50.37 µs  ▅█                  
                     (34.73 µs … 119.46 µs)  70.90 µs ▄███▅██▅▃▇▄▃▃▂▃▂▃▃▁▁▁
                  gc(  2.15 ms …   4.51 ms)   3.59 kb (  3.11 kb… 26.16 kb)
                   0.36 ipc ( 46.18% cache)  740.92 branch misses
        129.58k cycles  46.62k instructions   4.70k c-refs   2.53k c-misses

Rust WASM (small)              5.25 µs/iter   5.34 µs █▂▂█   ▆▂            
                        (4.97 µs … 5.92 µs)   5.90 µs ████▃▅▅██▇▁▅▁▁▃▃▃▃▁▁▃
                  gc(  2.24 ms …   5.43 ms)   3.11 kb (  3.09 kb…  3.52 kb)
                   3.27 ipc ( 69.45% cache)   17.71 branch misses
         18.94k cycles  61.92k instructions   67.10 c-refs   20.50 c-misses

Rust WASM (medium)             8.77 µs/iter   9.44 µs  █    ▅              
                       (7.14 µs … 12.38 µs)  11.45 µs ▇█▄▄▄▁█▇▇▇▄▇▄▄▄▁▁▄▁▁▄
                  gc(  2.41 ms …   6.44 ms) 218.68  b (214.03  b…364.67  b)
                   3.30 ipc ( 64.43% cache)   23.56 branch misses
         27.61k cycles  91.22k instructions   97.38 c-refs   34.64 c-misses

Rust WASM (large)            184.65 µs/iter 219.50 µs  ▇█▃                 
                    (128.17 µs … 358.25 µs) 332.92 µs ▃███▆▆▂▃▄▄▃▅▃▂▂▄▂▂▁▁▁
                  gc(  3.04 ms …   8.93 ms)  36.75 kb ( 29.88 kb… 77.92 kb)
                   1.75 ipc ( 35.92% cache)   1.68k branch misses
        397.69k cycles 695.75k instructions   6.54k c-refs   4.19k c-misses

Old js (warmup-ignore)        66.52 µs/iter  71.58 µs  ▃██▃▂▃              
                     (46.67 µs … 178.23 µs) 126.36 µs ▃██████▆▄▅▃▂▁▂▂▁▂▁▁▁▁
                  gc(  3.00 ms …   6.82 ms)   9.19 kb (  8.58 kb… 56.84 kb)
                   0.52 ipc ( 43.39% cache)  941.18 branch misses
        175.45k cycles  91.09k instructions   6.32k c-refs   3.58k c-misses

Old js (small)                14.14 µs/iter  14.43 µs   ▃     ▃ █   ▃      
                      (13.26 µs … 15.20 µs)  15.07 µs ▆▁█▆▁▆▁▆█▆█▆▁▆█▆▁▁▁▁▆
                  gc(  2.37 ms …   3.62 ms)   2.88 kb (  2.84 kb…  3.47 kb)
                   3.58 ipc ( 57.80% cache)   32.29 branch misses
         48.83k cycles 174.80k instructions  391.15 c-refs  165.06 c-misses

Old js (medium)               20.62 µs/iter  20.71 µs             █     █  
                      (20.34 µs … 20.77 µs)  20.75 µs █▁▁▁▁▁▁█▁▁▁███▁██▁█▁█
                  gc(  2.49 ms …   4.26 ms)   2.59 kb (  2.53 kb…  2.74 kb)
                   3.68 ipc ( 57.36% cache)   48.53 branch misses
         71.93k cycles 264.83k instructions  622.16 c-refs  265.31 c-misses

Old js (large)               371.65 µs/iter 401.43 µs   ▅█▃▄ ▇             
                    (264.36 µs … 750.28 µs) 665.32 µs ▄▇██████▆▆▂▂▁▂▂▁▂▁▁▁▂
                  gc(  2.11 ms …   9.01 ms) 358.00 kb (  2.30 kb…944.30 kb)
                   2.62 ipc ( 53.75% cache)   3.10k branch misses
          1.18M cycles   3.08M instructions  16.05k c-refs   7.42k c-misses
```

devlikepro commented:
So rust bridge is only about **2x** faster than the original JS version, right?
Not sure if adding Rust to the pure-js-typescript ecosystem is really worth it 🤔

Libsignal, for example, eats way more CPU time anyway.

purpshell commented:
> So rust bridge is only about **2x** faster than the original JS version, right? Not sure if adding Rust to the pure-js-typescript ecosystem is really worth it 🤔
> 
> Libsignal, for example, eats way more CPU time anyway.

We are planning to move libsignal under this rust bridge, and WASM allows for all runtime support. Also, the rust bridge is still *not optimized*. I see a promising future for this

vinikjkkj commented:
> i rewrited this using pure js and it's now x10 faster...

share with us

Santosl2 commented:
> i rewrited this using pure js and it's now x10 faster...

share with us

jlucaso1 commented:
I've made some optimizations (reducing allocations and removed some unecessary operations in the rust code).

Improved execution time and memory usage in operations.

New benchmark results ([bench.zip](https://github.com/user-attachments/files/22070877/bench.zip)):
```bash
➜  Baileys git:(feat-binary-node-rust-wasm) ✗ yarn tsx --expose-gc bench/bench.ts
clk: ~4.18 GHz
cpu: 12th Gen Intel(R) Core(TM) i5-12450H
runtime: node 22.18.0 (x64-linux)

benchmark                   avg (min … max) p75 / p99    (min … top 1%)
------------------------------------------- -------------------------------
• Encoding (JS Object -> Binary)
------------------------------------------- -------------------------------
Rust WASM (small)              4.92 µs/iter   4.95 µs          ▄█▅▄        
                        (4.68 µs … 5.17 µs)   5.14 µs ▃▁▁▁▃▆▄▆▄████▆▃▁▄▃▁▁▃
                  gc(  2.17 ms …   3.48 ms)   2.47 kb (  2.47 kb…  2.60 kb)
                   3.36 ipc ( 74.49% cache)   16.01 branch misses
         18.56k cycles  62.41k instructions   48.50 c-refs   12.37 c-misses

Rust WASM (medium)             7.28 µs/iter   7.31 µs    ▄ █     ▄▄ ▄█     
                        (7.20 µs … 7.36 µs)   7.35 µs ▅▅▅█▁█▅█▅█▅██▅██▁█▅▅▅
                  gc(  2.21 ms …   3.19 ms)   3.52 kb (  3.52 kb…  3.52 kb)
                   3.47 ipc ( 74.73% cache)   22.98 branch misses
         27.23k cycles  94.37k instructions   66.81 c-refs   16.89 c-misses

Rust WASM (large)             59.83 µs/iter  59.93 µs █       █      █     
                      (59.51 µs … 60.24 µs)  60.07 µs ██▁▁▁▁▁▁█▁██▁▁▁█▁▁▁██
                  gc(  2.28 ms …   3.21 ms)   1.01 kb (  1.01 kb…  1.01 kb)
                   3.61 ipc ( 75.05% cache)  254.49 branch misses
        222.48k cycles 802.85k instructions  542.33 c-refs  135.31 c-misses

Old js (small)                88.79 µs/iter  90.01 µs  █▇                  
                     (76.21 µs … 158.24 µs) 140.38 µs ▅███▅▃▁▂▂▃▄▃▂▂▂▁▁▁▁▁▁
                  gc(  2.10 ms …   4.12 ms)  27.64 kb ( 22.95 kb…119.06 kb)
                   1.14 ipc ( 56.55% cache)   1.53k branch misses
        271.81k cycles 309.33k instructions   7.89k c-refs   3.43k c-misses

Old js (medium)               19.43 µs/iter  19.46 µs             █        
                      (19.26 µs … 19.60 µs)  19.59 µs █▁▁▅▅▁▁▁▁▅▅▁█▁▁▁▁▅▁▁▅
                  gc(  2.39 ms …   4.97 ms)   1.96 kb (  1.88 kb…  2.21 kb)
                   3.68 ipc ( 59.53% cache)   56.31 branch misses
         71.58k cycles 263.50k instructions  607.44 c-refs  245.82 c-misses

Old js (large)               352.57 µs/iter 389.67 µs   ▂▃ ▆█▃ ▇▄▄▆▂▄ ▂▅   
                    (258.33 µs … 520.43 µs) 447.75 µs ▄▃██▅██████████▅██▆▄▄
                  gc(  2.32 ms …   8.13 ms) 391.87 kb ( 54.12 kb…  1.22 mb)
                   2.67 ipc ( 55.72% cache)   3.18k branch misses
          1.19M cycles   3.18M instructions  15.88k c-refs   7.03k c-misses
```

AstroX11 commented:
> i rewrited this using pure js and it's now x10 faster...

vague claims.

jlucaso1 commented:
Now everything is working (after the changes of LID stuffs, somethings has broken, but all has solved now).

The only missing on this PR is to improve the DX and reduce the breaking changes. Basically I've found that we can use rollup (bundler) in https://github.com/WhiskeySockets/whatsapp-rust-bridge that will drop the necessity of calling initWasm (async function in init of the scoket creation). The unique drawback is because the bundle of the library will become 30% larger (currently is about 180kb, btw the perfomance will not be changed).

https://github.com/rollup/plugins/tree/master/packages/wasm

vinikjkkj reviewed (approved):
Tested and working ✅

jlucaso1 committed:
perf(bridge): hand binary nodes over as one flat buffer

Both directions crossed the boundary per field. Decoding exposed a
handle whose every getter went back into WASM, and encoding let Rust
pull the node apart through Reflect; that was 56% of the encode profile,
against 12% doing the work. The handle path lost to the TypeScript
decoder by 9.9x on a 64-participant stanza.

decodeNodeFlat and encodeNodeFlat share one buffer format instead:

  u32 stringBytes, offsetCount, layoutCount, blobBytes
  string data | u32 offsets | u32 layout | blob data

Sections are 4-aligned so the reader takes views in place. Four things
made it fast, each from a profile rather than a guess:

- Tag and attribute strings are interned; converting them again on each
  occurrence was 13% of decode CPU.
- The builder is reused across calls; four allocations per decode showed
  up as allocator time.
- FxHash keyed on content, so interning does not allocate a String per
  distinct tag.
- Known tags travel as an index into the token table, which tokenTable()
  exposes once, rather than as bytes. This is what closed the gap on
  small stanzas, where fixed cost dominates.

Inflation is deliberately left to the caller: doing it here runs zlib
synchronously and stalls the loop long enough on a compressed group
stanza to miss deadlines.

jlucaso1 committed:
perf(binary): encode and decode through the bridge

Both directions now go through the flat buffer format, and the hand
written encoder, decoder and token tables come out: 1,600 lines less to
keep in step with the wire format.

Against the TypeScript they replace, on a group stanza with 1, 8 and 64
participants, interleaved in one process so drift hits both sides:

  encode   2.11x   2.80x   3.04x faster
  decode   1.05x   1.12x   1.20x faster

Encoder output is byte-identical, which the recorded parity vectors
check against the JS implementation. Decoded content comes back as a
Buffer rather than a bare Uint8Array, so callers doing
`(node.content as Buffer).toString('utf-8')` keep working: an earlier
handle-based attempt broke pairing on exactly that.

Inflation stays here rather than in the bridge, because node runs zlib
on the thread pool and the bridge would run it inline.

jlucaso1 committed:
bench(bridge): measure the stanzas the socket actually moves

The binary bench compared `encodeNode` and `decodeNode` against rc.9, but
Baileys stopped calling either: it goes through the flat codec. It also
timed `decodeNode` on its own, which returns a handle that materializes on
access, so the fastest line in the file measured a decode that had not
happened.

Bring in the shipped encoder and decoder by path and run them over an ack,
a one to one message, a device fanout at 8 and 64, and an app state patch,
compressed and not. The handle path keeps one line, against the flat path,
to show what it costs to read a whole tree through it.

binary-tree.ts goes: it existed to compare the handle and flat paths on a
group stanza, which this now covers.

jlucaso1 committed:
perf(binary): hand the decode over as borrowed sections

Returning the flat buffer as a `Uint8Array` cost around 500ns a call, near
enough fixed: a JS typed array to allocate, a wasm-bindgen heap slot to
take it through, and a fifth buffer to concatenate the four sections into
first. On a 31 byte ack that was most of the decode, and it is why the
smallest and most frequent stanzas lost to the TypeScript decoder while the
large ones won.

The decoder now leaves its sections where it wrote them and publishes their
offsets to a fixed address, which the caller reads out of linear memory. No
concatenation, no JS object per call, and the alignment the u32 views need
comes from the element type rather than from padding. What escapes the call
is byte content, so the decoder copies that section once and leaves the
leaves as views into the copy.

Three more, found by profiling what was left:

- FxHash walked a byte at a time and was 23% of the decode. Eight.
- `ValueRef::as_str` builds a `String` for every jid attribute, one per
  participant on a fanout. Format straight into the pool instead.
- Clearing the intern table cost the whole table, so a group stanza grew it
  and every small stanza after that paid to wipe it. Age entries by round.

On the JS side the reader state moves out of the closures the decode
captured, the encoder keeps its four collections across calls, the frame is
copied once rather than twice, and an uncompressed frame no longer takes a
second async hop to find out it needs no inflating.

Measured against baileys@7.0.0-rc.9, pinned to a P core:

  decode  ack 0.55us vs 0.69 | direct 1.96 vs 2.34 | fanout8 10.3 vs 12.9
          fanout64 62.7 vs 79.5 | app state 8.5 vs 12.0
  encode  ack 1.08us vs 1.15 | direct 1.73 vs 4.88 | fanout8 6.6 vs 22.0
          fanout64 40.9 vs 131.8 | app state 15.9 vs 129.3

The flat codec had no test of its own, only a bench, so it gets one: wire
parity against a recorded stanza, each content kind, the token boundary,
interning across rounds, and the two lifetime rules the borrow depends on.

jlucaso1 committed:
fix(binary): restore two wire behaviours the migration dropped

A jid with no user renders as a bare `s.whatsapp.net` in the core, and the
TypeScript decoder this replaced kept the `@`. Baileys routes on that:
`handleEncryptNotification` matches `from === S_WHATSAPP_NET`, so the
server's own pre-key count was falling through to the identity-change
branch and replenishment never ran. Two round-trip tests had recorded the
broken output as their expectation, and re-encoding a decoded frame now
reproduces it byte for byte again, matching rc.9 exactly.

The encoder also stringified attributes that were null or undefined,
putting the literal text `undefined` on the wire where the previous
encoder omitted the attribute. Optional attributes reach it unset from
real call sites, a USync user queried by phone and a media retry with no
participant among them. The count is backfilled rather than taken from a
filtered array, which would be an allocation per node and cost 15% of a
device fanout.

Two more, neither reachable from Baileys but both public:

- `encodeNodeFlat` derived its section spans with unchecked arithmetic, so
  a count whose byte length wraps on wasm32 passed the bounds check and
  then built a slice reaching far past the buffer. It also cast the byte
  spans to `*const u32`, which needs an alignment the caller's allocation
  does not promise. Both go: spans are checked, and words are read with
  `from_le_bytes`, which wasm does unaligned at no cost.
- `toJSON` walked the parsed node while the getters returned whatever a
  setter had put in their place, so a mutated handle serialized a stale
  tree. It now serializes what was written, and keeps the single-pass walk
  for a node nobody has written to.

jlucaso1 committed:
fix(binary): make toJSON see live writes, and guard the encoder again

`toJSON` walking the parsed node instead of the getters, which is how the
previous implementation read it, drops more than an assignment over
`node.attrs`. The object a getter hands back is live, and so is a child
handle, so `node.attrs.x = '...'` and a write to a child were both lost
while `encodeNode` and the getters themselves saw them.

Taking the caches whenever they are populated covers all three, and it
replaces the written flags from the previous commit, which only ever
caught assignment. A node nobody has touched still takes the single pass
over the parsed tree, which is every reader.

The encoder also lost two guards the TypeScript one had. An empty tag was
interned and shipped as a malformed stanza rather than throwing, and a
child left out of a conditionally built list arrives as null, which the
previous encoder dropped and this one crashed on while reading `tag` off
it. Both now behave as before, byte for byte against rc.9.

jlucaso1 committed:
chore(bridge): take the merged binary decoder changes

Picks up three that landed upstream: a byte-pair table for the packed
values, `smoothutf8` in place of `std` for validating what that table
just wrote, and dropping the per-node box around `NodeRef::content`.

The third one is breaking. `content` went from `Option<Box<NodeContentRef>>`
to `Option<NodeContentRef>`, so the three reads become `as_ref` and the
three sites in `FlatReader` stop boxing what they build. No logic moves.

Measured against the previous pin, alternating both builds three times
with a stanza that carries neither packed values nor content as a control
for drift: the ack is unchanged, a device fanout is around 5% faster. The
same commits measure well ahead of that natively, which is the third time
this decoder has gained less on wasm than on the host.

jlucaso1 committed:
bench(bridge): warm both decoders before timing anything

The first case in the file absorbed the JIT warm-up for the whole run,
worth around 150ns, and it landed on whichever stanza happened to be
first. That read as the smallest stanza being the one case the bridge
lost: 660ns against 620 for the TypeScript decoder. Moving the ack out of
first position put it at 497 against 695, and moved the penalty onto
whatever took its place.

With both decoders warmed over every fixture first, the ack decodes in
473ns against 572, and every other case keeps the standing it had.

jlucaso1 committed:
fix(binary): bound the intern table, which was leaking every string seen

The table that dedups strings within a decode aged its entries out by a
round counter instead of removing them, so every distinct message id, jid
and timestamp a socket ever decoded stayed in it. Two million decodes took
linear memory from 1.3 MB to 274 MB, growing in doublings as hashbrown
rehashed, and WASM never returns memory to the host. It is now flat at
1.3 MB across the same run.

Clearing the map each decode would also bound it, and that is what the
code did before; it was changed because clearing costs the whole table,
so one group stanza grew it and every small stanza after paid to wipe the
buckets left behind. A fixed 512-slot direct-mapped table has neither
problem: it cannot grow, it never needs clearing because the round stamp
still scopes hits to one decode, and a lookup is one indexed load instead
of a hashbrown probe.

Dedup was already best effort, which is what makes eviction on collision
sound: the caller writes the string twice and the pool gets longer. The
index bound in the hit check also covers the round counter wrapping after
four billion decodes onto a stale entry, which would have indexed out of
range.

Found by an audit agent reading the allocation behaviour, not by a
benchmark: the benchmarks decode the same three frames in a loop, so the
table stays at four entries and the growth is invisible to them.

jlucaso1 committed:
fix(binary): stop handing WASM memory to callers, and refuse a heap that hides growth

Two hazards in the borrowed-memory contract, found by an audit agent
reading the lifetime rules rather than by a test.

A stanza with no byte content leaves the blob section empty, and the
decoder stood in the whole-memory view for it. A node whose content was
an empty byte string therefore came back as a zero-length Buffer whose
`buffer` was all of linear memory at offset zero: nothing readable
through the view itself, but a live handle to the heap that Signal keys
live in, handed to application code. It now gets its own empty backing
store.

The cached views are rebuilt when a growth detaches them, which is
detected as a length of zero. That is only sound for a plain,
non-resizable ArrayBuffer: a SharedArrayBuffer never detaches, and a
resizable one is tracked by some view types and not others, so the check
would report "still valid" and the decoder would serve truncated strings
with no error. The runtime now refuses either at init.

The check reads the two properties directly instead of using
`instanceof`, which does not hold across realms: under jest's VM context
the host ArrayBuffer is not the context's, and the first version rejected
every buffer in the test suite.

jlucaso1 committed:
perf(socket): stop formatting a stack trace on every frame, and drop a dead hash

`promiseTimeout` captured `new Error().stack` and so did `delayCancellable`,
which it calls, so every outbound frame paid for two. Capturing the frames
costs about 1.2us; reading `.stack` makes V8 format them and costs about
11us. Worse, the formatting was not confined to the timeout path: settling
ran `cancel`, which built a Boom and read its stack, so a successful send
paid the full price. `promiseTimeout` only ever needed the timer dropped,
so `delayCancellable` grows a `clear` for that, and both errors now expose
the stack through a getter so the frames are formatted only if something
reads them.

That takes a `promiseTimeout` call from 22.0us to 3.0us. Every send goes
through one and every query through two.

`generateParticipantHashV2` was also being run per send and discarded:
`extraAttrs` was already spread into the `<enc>` nodes by the two awaited
`createParticipantNodes` calls above it, and nothing reads it afterwards.
Fixing the order would not have helped either, since the participant hash
belongs on the message stanza rather than on `<enc>`. Removed rather than
moved, because putting it on the wire is a protocol change and not a
performance one.

jlucaso1 committed:
fix(binary): keep the interop integrator, and ship the flat declaration

Two wire and packaging regressions from the migration.

An `INTEROP_JID` carries an integrator, and the core's `Display` leaves it
out: a jid with user `abc`, device 42 and integrator 7 came back as
`abc:42@interop` where the decoder this replaced wrote `7-abc:42@interop`.
Distinct interop identities collapse onto one string, which downstream
session and message routing keys on. Both the flat path and the handle now
go through one helper, so the two spellings cannot drift; the server-only
`@` case moved into it as well.

`ts/index.ts` re-exports `./flat.js`, so the generated `dist/index.d.ts`
references `dist/flat.d.ts`, which the `files` list did not carry. The
failure is quiet: esbuild inlines the module so the runtime works, and
TypeScript drops the unresolvable re-export and reports `decodeNodeFlat`
and `encodeNodeFlat` as simply not exported, suggesting the raw
`__decodeNodeFlat` instead. Confirmed against a real `npm pack`.

`check-package` now walks the relative re-exports of every packed
declaration and fails when one does not resolve inside the tarball, so the
class of bug cannot come back for the next file.
